### ISSUE TITLE
High-tier potion bonus spells on tier-checked wands

### ISSUE THREAD

**Issue by markuseerikhorma** [TheHorscht/Anvil-of-Destiny#7]
Issue: 
Some potion and item bonuses can include spells exclusive to tier 10 or with average rounded up level of 8 or 9¹. 
[Potion bonuses](https://github.com/TheHorscht/Anvil-of-Destiny/blob/d5d7d2f8e3850073dec633b451f54b5631115de3/files/entities/anvil/potion_bonuses.lua#L409), [used by anvil](https://github.com/TheHorscht/Anvil-of-Destiny/blob/master/files/entities/anvil/anvil.lua#L226C1-L226C1 ), sometimes checks those cases against wand tier ([Wand Utils add spells to wand](https://github.com/TheHorscht/Anvil-of-Destiny/blob/d5d7d2f8e3850073dec633b451f54b5631115de3/files/entities/anvil/wand_utils.lua#L223) > [wand compute level](https://github.com/TheHorscht/Anvil-of-Destiny/blob/d5d7d2f8e3850073dec633b451f54b5631115de3/files/entities/anvil/wand_utils.lua#L17)) , with few exceptions³, which is [capped at 6](https://github.com/TheHorscht/Anvil-of-Destiny/blob/d5d7d2f8e3850073dec633b451f54b5631115de3/files/entities/anvil/wand_utils.lua#L17).

Testing procedure:

Version: 2021 version of Noita and version 1.19.0 of Anvil of Destiny retrieved from modworkshop.

Settings: 
Anvil of Destiny default configuration modified with Lava lake anvil and reusable anvil.
Play Last Seed to force starting with seeds 71772 & 90345 (mud/soil/water Alchemical Precursor, second with rain)
No red screen on low HP (test 1). 

Test 1, seed 1: 
1. Make 100% full flask of midas.
2. Find an All-Seeing Eye mines wand to the left with 6 capacity and drop it to lava lake anvil. 
3. Use Interact key with the full flask of midas. 

Result: 3 Touch of Gold spells were added to the wand.

Test 2, seed 2: 
1. Make 100% full flask of midas.
2. Drop starting wand with 1 spark bolt, 3 capacity to lava lake anvil.
3. Use Interact key with the full flask of midas. 

Result: +1 capacity and 2 Touch of Gold spells were added to the wand. 

Test 3, seed 2:
4. Fill flask back up with Draught of Midas
5. Purchase and empty 9 capacity wand in first holy mountain with 480 max mana / 11 mana recharge
6. Find an anvil in coal pits to drop said wand on, then interact with potion.

Result: 2 Capacity and 4 Touch of Gold spells were added to the wand.

Test 4, seed 2:
7. Restart by suicide & new game
8. Drop starting wand with 1 spark bolt, 3 capacity to lava lake lanvil.
9. Use interact key with full flask of Chaotic Polymorphine

Result: 1 Tentacler spell was added to the wand.

(not specified: exact values of buffs to wands' mana stats, though they were present.). 

Expected behavior: 
a) Tier 7 spells, like Touch of Gold, added only to tier 6 wands found in Temple of the Art and elsewhere. 
b) Tier 10 spells are never added.
c) Therefore, no spells added to starting/mines/first holy mountain wands.

Actual behavior: 
Touch of Gold was added 9 times between the three tests. End of Everything was added 0 times between three tests. Tentacler was added, and it is not the lowest tier spell available for chaotic poly bonus (summon egg is lower by both average and minimum).

Given that, looks like some kind of checking is present, but defaults to always giving at least some spells? Exclusively touch of gold results suggests lowest tier spells, but the Tentacler test goes to disprove that idea.

1 midas (ALL_SPELLS), wandstone (RESET, DIVIDE_10, others), sunseed/sunstone (BOMB_HOLY_GIGA), berserkium/gunpowder_unstable (ALL_NUKES)
3 Gold(L137), Monstrous Powder (L152), Healthium & Lively Concoction (L200), Healthy Porridge(L361), Gourd (L581), Kammi(L613), Egg (L624), Sädekivi (L673), Kuulokivi (L698), Shiny Orb (L707), Broken Portal (L754), Crystal Key (L774), Kakkakikkare (L783) in potion_bonuses.lua

**Comment by TheHorscht** [comment]
Yes I see the problem, I never took the `spawn_probability`s of spells into account, only `spawn_level`.
So for `TOUCH_GOLD` it is:
```
spawn_level       = "1,2,3,4,5,6,7,10"
spawn_probability = "0,0,0,0,0.1,0.1,0.1,0.5"
```
which should make it ignore levels 1,2,3,4,5. I wonder why they did it this way instead of just not including those levels in the `spawn_level`s. The problem is that no wand level will ever be higher than 6, so if I somehow took spell probabilities into account I would have to somehow merge anything above level 6 into level 6 I think. Not sure what the best way to solve this would be. Maybe some custom logic in addition to incorporating spell levels? Because you probably should be getting touch of gold at least sometimes even on lower level wands, for such a high rarity potion.

**Comment by markuseerikhorma** [comment]
I see. I saw the spawn probability being ignored, but wrongly assumed wiki was accurate down to the unimportant technicalities. I guess it does note about 19 Dec 2019: 
>       All Touch of spells removed from tier 10 spells.
>        Added all Touch of Spells to tiers 1 through 7, all with spawn rates of 0.1

(iirc, Note spells are only ones where appearance probability of 0 is noted.)


If you ignored zero-probability tiers for level calculation while otherwise keeping it as-is, Touch of Spells would be treated as 7; barely capable of appearing on tier 6 wands. At my glance, Spells to nukes is only non-pure T10 spell that doesn't become obtainable in this way in potion_bonuses.

As for getting at least some spells sometimes...Doesn't compare that well in cases where lower tier spell lists aren't empty. I don't know what solution would fit your vision (Concentrate lower-tier spell slots into lesser number of higher-tier ones in case of limited space? Convert one of the stats into % chance? (Though Mana Recharge and Capacity are already most important stats on wands.)), but if you cannot decide you could default to the vanilla solution of recursively having 10% chance to roll a spell of higher tier. 

Granted, vanilla method would give similar odds for Touch of Gold on starter wand as Giga Nuke off Summon Taikasauva, but you could get lucky with at least 1 on good capacity tier 5 wand.


Initial expected solution was to ignore wand tiers on bonuses with T10 spells, but that was when I had just read the code and had not actually enable Anvil of Destiny in noita once yet.

But of course, there are gameplay advantages with having spells limited to particular tier, as it'd allow picking a wand to limit your spell choice, so might not want to go full (weighted) random result either. 


PS: Side tip for getting full flask of midas: you need a large pool of alchemic precursor first, because once you convert it to midas you often cannot make more midas in same pool because midas will convert solids you'd use for precursor into gold. So you need to make additional midas somewhere else and carry a drop over if you don't have a convenient corpse.

**Comment by TheHorscht** [comment]
Should be fixed now in 789e5ef, although I didn't rewrite it entirely to take the spell probabilities into account for the recipes, since it would mess up everything too much, especially the expectations of people who got used to the old system. It was just a couple of spells, mostly touch of X and some ray spells. They should no longer appear on lower level wands.
